### PR TITLE
Add source span annotations to Declaration

### PR DESCRIPTION
**Pull request by garyb**
Approx 1,000,000 things left to re-examine still. I got bored dealing with `Parser` hence the `undefined`s for now :stuck_out_tongue:

### CODE REVIEW

**Comment by garyb** [comment]
I decided to go from the top this time, with the `Declaration`s, as it should be possible to eliminate the non-`SourceSpan`-accepting `errorMessage` right now. We'll always be guaranteed to have at least a vaguely local source span available if `Declaration`s have it.

File: tests/Language/PureScript/Ide/SourceFileSpec.hs [hunk 1]
```diff
@@ -16,55 +16,62 @@ span0 = P.SourceSpan "ModuleLevel" (P.SourcePos 0 0) (P.SourcePos 1 1)
...
-member1 = P.TypeDeclaration (P.Ident "member1") P.REmpty
+foreign1 = P.ExternDeclaration ann1 (P.Ident "foreign1") P.REmpty
+foreign2 = P.ExternDataDeclaration ann1 (P.ProperName "Foreign2") P.kindType
+foreign3 = P.ExternKindDeclaration ann1 (P.ProperName "Foreign3")
+member1 = P.TypeDeclaration ann2 (P.Ident "member1") P.REmpty
 
 spec :: Spec
 spec = do
   describe "Extracting Spans" $ do
+    --TODO-gb: is this worth testing now? the spans part not so much, but returning the appropriate IdeNSValue/IdeNSType is useful perhaps?
```
Comment (garyb) [inline tests/Language/PureScript/Ide/SourceFileSpec.hs]: What say you @kRITZCREEK?

Comment (kRITZCREEK) [inline tests/Language/PureScript/Ide/SourceFileSpec.hs]: These tests have caught the stupidest mistakes before, so I'll say they're worth it. At the very least they make sure all the pattern matches are exhaustive.

**Comment by garyb** [comment]
This is ready for review now.

I've not fully eliminated the old `errorMessage` yet - the remaning cases could be fixed now, but would involve some work that will just be undone again in the next PR, which will be for annotating expressions/binders.

I think there may be two cases now that are potentially unpositioned, as I removed a `rethrow`, but those will be fixed again anyway once `errorMessage` is done.

**Comment by paf31** [comment]
👍 Fantastic, thanks!

**Pull request merged**